modules/drivers/snowflake/src/metabase/driver/snowflake.clj

```diff
@@ -227,69 +227,57 @@
-                              (hsql/call
-                               :or
-                               (hsql/call :> (extract :month a) (extract :month b))
-                               (hsql/call
-                                :and
-                                (hsql/call := (extract :month a) (extract :month b))
-                                (hsql/call :> (extract :day a) (extract :day b)))))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :quarter
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx//
-                               (hx/-
-                                (hsql/call :datediff (hsql/raw "month") a b)
-                                (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b))))
-                               3))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :month
-      (let [positive-diff (fn [a b]
-                            (hx/-
-                             (hsql/call :datediff (hsql/raw "month") a b)
-                             (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b)))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :week
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (hsql/call :datediff (hsql/raw "day") a b) 7))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :day
-      (hsql/call :datediff (hsql/raw (name unit)) x y)
-
-      (:hour :minute :second)
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (hx/cast :float (hsql/call :datediff (hsql/raw "millisecond") a b)) ; datediff returns integer, so cast to float
-                                    (case unit :hour 3600000 :minute 60000 :second 1000)))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x)))))))
+(defn- time-zone-aware-datediff
+  "Same as snowflake's `datediff` but accounts for the results time zone when calculating date boundaries.
+   This is needed when an argument is of timestamptz type and the unit is day, week, month, quarter or year."
+  [unit x y]
+  (let [x (cond->> x
+            (hx/is-of-type? x "timestamptz")
+            (hsql/call :convert_timezone (qp.timezone/results-timezone-id)))
+        y (cond->> y
+            (hx/is-of-type? y "timestamptz")
+            (hsql/call :convert_timezone (qp.timezone/results-timezone-id)))]
```

Comment:
I added this function because it's not necessary to use `convert_timezone` if the arg is not a timestamptz, and the unit is second, minute or hour.